src/bcd/bcd/MultiscaleDenoiser.cpp

```diff
@@ -29,6 +29,62 @@ using namespace std;
 
 namespace bcd
 {
+namespace
+{
+
+class MultiScaleProgressReporter
+  : public IProgressReporter
+{
+  public:
+    MultiScaleProgressReporter(
+        int                 i_scale,
+        int                 i_nbOfScales,
+        IProgressReporter*  i_progress_reporter)
+      : m_scale(i_scale)
+      , m_nbOfScales(i_nbOfScales)
+      , m_progress_reporter(i_progress_reporter)
+    {
+    }
+
+    void progress(const float i_progress) const override
+    {
+        if (m_progress_reporter)
+        {
+            if (m_scale == m_nbOfScales - 1)
+            {
+                m_progress_reporter->progress(
+                    i_progress / float(((1 << (2 * m_nbOfScales)) - 1) / 3));
+            }
+            else
+            {
+                int s = m_nbOfScales - 1 - m_scale;
+                // next higher definition scale is 4 times slower
+                // 1 + 4 + 4^2 + ... 4^s = (4^(s+1) - 1) / (4 - 1) = (2^(2*(s+1)) - 1) / 3
+                // = ((1 << (2*(s+1))) - 1) / 3
+                float factor = 1.f / float(((1 << (2 * m_nbOfScales)) - 1) / 3);
```

Comment:
May I suggest to move `float(((1 << (2 * m_nbOfScales)) - 1) / 3)` to a free function and call it four times here?

Reply:
Make const.